From 78312420395e0f31849e6cb66db1aa62c3d88b65 Mon Sep 17 00:00:00 2001
From: Myron Stowe <myron.stowe@redhat.com>
Date: Fri, 5 Sep 2014 02:16:20 -0400
Subject: [pci] Add bridge DMA alias quirk for ASMedia and Tundra bridges

Message-id: <20140905021620.32574.96413.stgit@gir.stowe>
Patchwork-id: 91366
O-Subject: [RHEL7.1 PATCH 07/17] PCI: Add bridge DMA alias quirk for ASMedia and Tundra bridges
Bugzilla: 1136537
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Stefan Assmann <sassmann@redhat.com>
RH-Acked-by: Alex Williamson <alex.williamson@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

Bugzilla: https://bugzilla.redhat.com/show_bug.cgi?id=1136537
Upstream Status: ebdb51eb7806df56221d4536a6a9702751d6df85

commit ebdb51eb7806df56221d4536a6a9702751d6df85
Author: Alex Williamson <alex.williamson@redhat.com>
Date:   Thu May 22 17:08:07 2014 -0600

    PCI: Add bridge DMA alias quirk for ASMedia and Tundra bridges

    The quirk is intended to be extremely generic, but we only apply it to
    known offending devices.

    Link: https://bugzilla.kernel.org/show_bug.cgi?id=44881
    Signed-off-by: Alex Williamson <alex.williamson@redhat.com>
    Signed-off-by: Bjorn Helgaas <bhelgaas@google.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/pci/quirks.c b/drivers/pci/quirks.c
index 781111d..5170642 100644
--- a/drivers/pci/quirks.c
+++ b/drivers/pci/quirks.c
@@ -3391,6 +3391,29 @@ DECLARE_PCI_FIXUP_HEADER(PCI_VENDOR_ID_JMICRON,
     PCI_DEVICE_ID_JMICRON_JMB388_ESD,
     quirk_dma_func1_alias);
 
+/*
+ * A few PCIe-to-PCI bridges fail to expose a PCIe capability, resulting in
+ * using the wrong DMA alias for the device.  Some of these devices can be
+ * used as either forward or reverse bridges, so we need to test whether the
+ * device is operating in the correct mode.  We could probably apply this
+ * quirk to PCI_ANY_ID, but for now we'll just use known offenders.  The test
+ * is for a non-root, non-PCIe bridge where the upstream device is PCIe and
+ * is not a PCIe-to-PCI bridge, then @pdev is actually a PCIe-to-PCI bridge.
+ */
+static void quirk_use_pcie_bridge_dma_alias(struct pci_dev *pdev)
+{
+ if (!pci_is_root_bus(pdev->bus) &&
+     pdev->hdr_type == PCI_HEADER_TYPE_BRIDGE &&
+     !pci_is_pcie(pdev) && pci_is_pcie(pdev->bus->self) &&
+     pci_pcie_type(pdev->bus->self) != PCI_EXP_TYPE_PCI_BRIDGE)
+  pdev->dev_flags |= PCI_DEV_FLAG_PCIE_BRIDGE_ALIAS;
+}
+/* ASM1083/1085, https://bugzilla.kernel.org/show_bug.cgi?id=44881#c46 */
+DECLARE_PCI_FIXUP_HEADER(PCI_VENDOR_ID_ASMEDIA, 0x1080,
+    quirk_use_pcie_bridge_dma_alias);
+/* Tundra 8113, https://bugzilla.kernel.org/show_bug.cgi?id=44881#c43 */
+DECLARE_PCI_FIXUP_HEADER(0x10e3, 0x8113, quirk_use_pcie_bridge_dma_alias);
+
 static struct pci_dev *pci_func_0_dma_source(struct pci_dev *dev)
 {
  if (!PCI_FUNC(dev->devfn))
-- 
1.7.1